synapse/module_api/__init__.py

```diff
@@ -700,6 +715,38 @@ def read_templates(
             (td for td in (self.custom_template_dir, custom_template_directory) if td),
         )
 
+    def ts(self) -> int:
+        """Return a timestamp for the current time, in milliseconds since the epoch"""
+        return self._hs.get_clock().time_msec()
+
+    def is_mine(self, domain_specific_string: DomainSpecificString) -> bool:
+        """Checks whether an ID comes from this homeserver."""
+        return self._hs.is_mine(domain_specific_string)
+
+    def is_mine_id(self, string: str) -> bool:
+        """Checks whether an ID comes from this homeserver."""
+        return string.split(":", 1)[1] == self._server_name
+
+    async def get_user_ip_and_agents(
+        self, user: UserID, since_ts: Optional[int] = None
+    ) -> List[UserIpAndAgent]:
+        """
+        Return the list of user IPs and agents for a user.
+
+        Only useful for local users.
+        """
+        raw_data = await self._store.get_user_ip_and_agents(user)
+        # Sanitize some of the data. We don't want to return tokens.
+        return [
+            {
+                "ip": str(data["ip"]),
+                "user_agent": str(data["user_agent"]),
+                "last_seen": int(data["last_seen"]),
+            }
+            for data in raw_data
+            if since_ts is None or int(data["last_seen"]) >= since_ts
+        ]
```

Comment:
This returns a list of dicts, not a list of `UserIpAndAgent`

Reply:
Almost sure it does :)

Reply:
Well sure it does now :p

Reply:
I'm almost sure that you can coerce a dictionary into a subclass of `TypedDict`.